lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,517 @@
+    break;
+  case BTOR2_TAG_uext:
+    res = builder.create<btor::UExtOp>(
+        unknownLoc, cache.at(kids[0]),
+        builder.getIntegerType(line->sort.bitvec.width));
+    break;
+
+  // ternary ops
+  case BTOR2_TAG_ite:
+    res = builder.create<btor::IteOp>(unknownLoc, cache.at(kids[0]),
+                                      cache.at(kids[1]), cache.at(kids[2]));
+    break;
+
+  // unmapped ops
+  case BTOR2_TAG_read:
+  case BTOR2_TAG_constraint:
+  case BTOR2_TAG_init:
+  case BTOR2_TAG_input:
+  case BTOR2_TAG_next:
+  case BTOR2_TAG_sort:
+  case BTOR2_TAG_state:
+  case BTOR2_TAG_fair:
+  case BTOR2_TAG_justice:
+  case BTOR2_TAG_output:
+  case BTOR2_TAG_write:
+  default:
+    break;
+  }
+  return res;
+}
+
+void Deserialize::createNegateLine(int64_t curAt, Value &child) {
+  auto res = builder.create<btor::NotOp>(unknownLoc, cache.at(curAt * -1));
+  assert(res && res->getNumResults() == 1);
+  cache[curAt] = res->getResult(0);
+}
+
+bool Deserialize::isValidChild(Btor2Line * line) {
+  auto tag = reachedLines.at(line->id)->tag;
+  if (tag == BTOR2_TAG_init || tag == BTOR2_TAG_constraint ||
+      tag == BTOR2_TAG_next || tag == BTOR2_TAG_read ||
+      tag == BTOR2_TAG_state || tag == BTOR2_TAG_input ||
+      tag == BTOR2_TAG_sort || tag == BTOR2_TAG_fair ||
+      tag == BTOR2_TAG_justice || tag == BTOR2_TAG_output ||
+      tag == BTOR2_TAG_write) {
+    return false;
+  }
+  return true;
+}
+
+///===----------------------------------------------------------------------===//
+/// This function's goal is to add the MLIR Operation that corresponds to 
+/// the given Btor2Line* into the basic block designated by the provided 
+/// builder. Then, the MLIR Value of the newly minted operation is added
+/// into our cache for future reference within the basic block. 
+///
+/// e.x:
+///      for (auto it = nexts.begin(); it != nexts.end(); ++it) {
+///          toOp(*it);
+///      }
+///
+///  We can see that for each next operation in btor2, we will compute all
+///  the prerequisite operations before storing the result in our cache
+///===----------------------------------------------------------------------===//
+void Deserialize::toOp(Btor2Line *line) {
+  if (cache.find(line->id) != cache.end()) {
+    return;
+  }
+
+  Operation *res = nullptr;
+  std::vector<Btor2Line *> todo;
+  todo.push_back(line);
+  while (!todo.empty()) {
+    auto cur = todo.back();
+    uint32_t oldsize = todo.size();
+    for (uint32_t i = 0; i < cur->nargs; ++i) {
+      if (cur->args[i] > 0 
+      && !isValidChild(reachedLines.at(cur->args[i]))) {
+        continue;
+      }
+
+      if (cache.find(cur->args[i]) == cache.end()) {
+        if (cur->args[i] < 0) {
+          // if original operation is cached, negate it
+          if (cache.find(cur->args[i] * -1) != cache.end()) {
+            createNegateLine(cur->args[i], cache.at(cur->args[i] * -1)); 
+          } else {
+            todo.push_back(reachedLines.at(cur->args[i] * -1));
```

Comment:
I would use `std::abs` instead of multiplication by `-1` since it is not clear whether the result of multiplication is expected to be positive. This can be gotten from context, but it is not obvious.

Check that `std::abs` works with integers without converting them to floats

Reply:
I don't see an indication that `std::abs` converts integers to floats in the cpp reference. Will go ahead an use it.

Reply:
and*